src/metabase/pulse/render/datetime.clj

```diff
@@ -11,33 +11,107 @@
            java.time.Period
            java.time.temporal.Temporal))
 
+(defn temporal-string?
+  "Returns `true` if the string `s` is parseable as a datetime.
+
+  `(temporal-string? \"asdf\")` -> false
+  `(temporal-string? \"2020-02-02\")` -> true"
+  [s]
+  (boolean
+    (try
+      (u.date/parse s)
+      (catch Exception _e false))))
+
 (defn- reformat-temporal-str [timezone-id s new-format-string]
   (t/format new-format-string (u.date/parse s timezone-id)))
 
+(defn- day-of-week
+  [s abbreviate]
+  (get-in
+    {"1" {:full (tru "Sunday") :abbreviated (tru "Sun")}
+     "2" {:full (tru "Monday") :abbreviated (tru "Mon")}
+     "3" {:full (tru "Tuesday") :abbreviated (tru "Tue")}
+     "4" {:full (tru "Wednesday") :abbreviated (tru "Wed")}
+     "5" {:full (tru "Thursday") :abbreviated (tru "Thu")}
+     "6" {:full (tru "Friday") :abbreviated (tru "Fri")}
+     "7" {:full (tru "Saturday") :abbreviated (tru "Sat")}}
+    (if abbreviate
+      [s :abbreviated]
+      [s :full])))
+
+(defn- month-of-year
+  [s abbreviate]
+  (get-in
+    {"1"  {:full (tru "January") :abbreviated (tru "Jan")}
+     "2"  {:full (tru "February") :abbreviated (tru "Feb")}
+     "3"  {:full (tru "March") :abbreviated (tru "Mar")}
+     "4"  {:full (tru "April") :abbreviated (tru "Apr")}
+     "5"  {:full (tru "May") :abbreviated (tru "May")}
+     "6"  {:full (tru "June") :abbreviated (tru "Jun")}
+     "7"  {:full (tru "July") :abbreviated (tru "Jul")}
+     "8"  {:full (tru "August") :abbreviated (tru "Aug")}
+     "9"  {:full (tru "September") :abbreviated (tru "Sep")}
+     "10" {:full (tru "October") :abbreviated (tru "Oct")}
+     "11" {:full (tru "November") :abbreviated (tru "Nov")}
+     "12" {:full (tru "December") :abbreviated (tru "Dec")}}
+    (if abbreviate
+      [s :abbreviated]
+      [s :full])))
+
+(defn- x-of-y
+  "Format an integer as x-th of y, for example, 2nd week of year."
```

Comment:
I don't know if we want to pull this dep in but it could be useful:

```clojure
clj -Sdeps '{:deps {com.ibm.icu/icu4j {:mvn/version "71.1"}}}'
user=> (import 'java.util.Locale 'com.ibm.icu.text.RuleBasedNumberFormat)
com.ibm.icu.text.RuleBasedNumberFormat
user=> (doseq [[locale n] [[Locale/GERMAN "german"]
                    [Locale/JAPANESE "japanese"]
                    [Locale/ITALY "italian"]
                    [Locale/FRANCE "french"]
                    [Locale/US "american"]]]
  (let [nf (RuleBasedNumberFormat. ^Locale locale RuleBasedNumberFormat/ORDINAL)]
    (println n ":")
    (doseq [i [1 2 3 4 10 11 1000 1001 1011]]
      (println i (.format nf i "%digits-ordinal")))
    (println)))
german :
1 1.
2 2.
3 3.
4 4.
10 10.
11 11.
1000 1.000.
1001 1.001.
1011 1.011.

japanese :
1 第1
2 第2
3 第3
4 第4
10 第10
11 第11
1000 第1,000
1001 第1,001
1011 第1,011

italian :
1 1º
2 2º
3 3º
4 4º
10 10º
11 11º
1000 1.000º
1001 1.001º
1011 1.011º

french :
1 1er
2 2e
3 3e
4 4e
10 10e
11 11e
1000 1 000e
1001 1 001e
1011 1 011e

american :
1 1st
2 2nd
3 3rd
4 4th
10 10th
11 11th
1000 1,000th
1001 1,001st
1011 1,011th

nil
```

Reply:
Probably a good idea... especially considering I didn't account for 12 -> "12nd"... if I miss things in my own language, I think I can't be sure things will work properly with other locales either

Reply:
I think that lib is pulled in with another of our dependencies. I was able to import it without changing anything in the deps.edn.

Reply:
bizarre. It seems to be vendored in from graal?
```clojure
(clojure.java.io/resource "com/ibm/icu/text/RuleBasedNumberFormat.class")
#object[java.net.URL
        "0x16eae632"
        "jar:file:/Users/dan/.m2/repository/org/graalvm/regex/regex/22.1.0/regex-22.1.0.jar!/com/ibm/icu/text/RuleBasedNumberFormat.class"]
```

That seems to be in the `22.1.0` jar, but not in 20.1.0, 21.2.0, or 22.0.0.2. I think we need to add this version to our top level dependencies so we don't break if they remove this in the future. I don't understand why they version it like this though.

Reply:
If we didn't have this and didn't want to pull in another dep for some reason, I think `clojure.pprint/cl-format` can do this too. But I think it's only for English